internal/httpapi/httpapi.go

```diff
@@ -0,0 +1,80 @@
+// Copyright (c) 2021 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package httpapi
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/gorilla/mux"
+
+	"github.com/canonical/pebble/internal/logger"
+	"github.com/canonical/pebble/internal/overlord/checkstate"
+)
+
+type API struct {
+	checkMgr CheckManager
+	router   *mux.Router
+}
+
+type CheckManager interface {
+	Checks() ([]*checkstate.CheckInfo, error)
+}
+
+func NewAPI(checkMgr CheckManager) *API {
+	s := &API{
+		checkMgr: checkMgr,
+		router:   mux.NewRouter(),
+	}
+
+	s.router.NotFoundHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		writeError(w, http.StatusNotFound, "not found")
+	})
+	s.router.MethodNotAllowedHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		writeError(w, http.StatusMethodNotAllowed, "method not allowed")
+	})
+
+	s.router.HandleFunc("/v1/health", s.getHealth).Methods("GET")
```

Comment:
Similar idea here: this is being bolted on ignoring everything else we already have in Pebble. Note that the stock daemon in Pebble already supports multiple sockets, where each socket supports different endpoints, and potentially even different parameters depending on how the actual handshake with the protocol was made.

It may be useful to understand how this is used by snapd: we have two different sockets, one that is trusted, and one that is untrusted. The untrusted socket will never be able to access certain priviliged APIs, while the trusted socket may access certain APIs, but only if it's being accessed by root. In other words, the untrusted socket cannot see the privileged APIs even if it is accessing the socket as root, because inside the container being root is not actually a sign of privilege as far as snapd is concerned.

So, coming back to our use case here, this is super useful isn't it?  We have exactly the same scenario: one socket that is privileged, and one socket that is open and unprivileged. So we should be able to use exactly the same mechanics already in place to dispatch requests, instead of having this completely out of band API.

Makes sense?

Reply:
Good call -- I hadn't looked at that closely enough. I moved it from a totally separate `httpapi` into the regular `daemon` package and used the existing `GuestOK` auth field -- meaning world-accessible. The "guest API" is what's exposed via `pebble -http=<address>` now. This made things quite a lot simpler as it reused a bunch of the existing code (-150 lines or so).